[aotd] non-contiguous NestedTensor mutation in compile

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139630



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"

IvanKobzarev committed:
[WIP][aotd] non-contiguous NestedTensor mutaiton in compile

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1891,15 +1891,63 @@ def _backward_prologue(ctx, *flat_args):
+                            ],
+                        ):
+                            if isinstance(m, SubclassCreationMeta):
+                                if type(t) == torch.Tensor:
```
I think this is technically wrong: this assumes that if we guessed the tangent was a subclass, and it was actually a plain tensor (and it maps to a mutated fw input), then it **must** be a tensor of zeros. But that does not have to be true: it's possible that someone later used the (updated) input in some autograd-related compute, but the subclass disappeared when running the backward.

Rather than making a guess about adding extra zero tensor arguments, I think we should either:

(1) "guess better": today, we make the guess when generating our FakeTensor tangents for tracing that if any of the fw_outs are subclasses, their tangents will also be subclasses. At the time that we generate this guess, we could tweak it so that if a `fw_out` corresopnds to an `updated_input` that we are returning, then we can guess that it will be a plain tensor

(2) if this is too much of a pain, we can also just give up on this case for now

IvanKobzarev replied:
Thanks! I agree, guessing better is much cleaner. updated the PR with this logic.

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

IvanKobzarev commented on torch/_functorch/_aot_autograd/functional_utils.py:
```diff
@@ -457,9 +457,11 @@ def assert_functional_graph(fx_g: torch.fx.Graph) -> int:
-                    assert (
-                        n.args[0] in placeholders
-                    ), f"n={str(n)}, n.args[0]={str(n.args[0])}, placeholders={str(placeholders)}, graph={str(fx_g)}"
+                    tensor_meta = n.args[0].meta["tensor_meta"]
```
I could be wrong, lifting this assert.
We get the graph with copy_(arg.detach(), ...slice)
when compile NestedTensor index_put
```
python test/test_nestedtensor.py -k non_contiguous_mutation 
```
```
def fn(x, x0):
    x[0, 0, 0] = 2
    return x
```

for inputs torch.nested.nested_tensor_from_jagged(), torch.ones(2, 32)

Results in  a graph with mutation:
```
AssertionError: n=copy_, n.args[0]=detach_3, placeholders={arg4_1, arg3_1, arg1_1, arg2_1, arg0_1}, 

graph=graph():
    %arg0_1 : [num_users=0] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=0] = placeholder[target=arg1_1]
    %arg2_1 : [num_users=4] = placeholder[target=arg2_1]
    %arg3_1 : [num_users=3] = placeholder[target=arg3_1]
    %arg4_1 : [num_users=1] = placeholder[target=arg4_1]
    %_tensor_constant0 : [num_users=1] = get_attr[target=_tensor_constant0]
    %lift_fresh_copy : [num_users=1] = call_function[target=torch.ops.aten.lift_fresh_copy.default](args = (%_tensor_constant0,), kwargs = {})
    %detach : [num_users=1] = call_function[target=torch.ops.aten.detach.default](args = (%arg2_1,), kwargs = {})
    %slice_1 : [num_users=1] = call_function[target=torch.ops.aten.slice.Tensor](args = (%arg3_1, -1, 0, 2), kwargs = {})
    %slice_2 : [num_users=1] = call_function[target=torch.ops.aten.slice.Tensor](args = (%arg3_1, -1, 1, 3), kwargs = {})
    %sub : [num_users=1] = call_function[target=torch.ops.aten.sub.Tensor](args = (%slice_2, %slice_1), kwargs = {})
    %unbind : [num_users=2] = call_function[target=torch.ops.aten.unbind.int](args = (%sub,), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%unbind, 0), kwargs = {})
    %getitem_1 : [num_users=1] = call_function[target=operator.getitem](args = (%unbind, 1), kwargs = {})
    %_local_scalar_dense : [num_users=4] = call_function[target=torch.ops.aten._local_scalar_dense.default](args = (%getitem,), kwargs = {})
    %_local_scalar_dense_1 : [num_users=3] = call_function[target=torch.ops.aten._local_scalar_dense.default](args = (%getitem_1,), kwargs = {})
    %split_with_sizes : [num_users=2] = call_function[target=torch.ops.aten.split_with_sizes.default](args = (%detach, [%_local_scalar_dense, %_local_scalar_dense_1]), kwargs = {})
    %getitem_2 : [num_users=1] = call_function[target=operator.getitem](args = (%split_with_sizes, 0), kwargs = {})
    %getitem_3 : [num_users=0] = call_function[target=operator.getitem](args = (%split_with_sizes, 1), kwargs = {})
    %select : [num_users=1] = call_function[target=torch.ops.aten.select.int](args = (%getitem_2, 0, 0), kwargs = {})
    %select_1 : [num_users=1] = call_function[target=torch.ops.aten.select.int](args = (%select, 0, 0), kwargs = {})
    %copy : [num_users=1] = call_function[target=torch.ops.aten.copy.default](args = (%select_1, %lift_fresh_copy), kwargs = {})
    %detach_1 : [num_users=2] = call_function[target=torch.ops.aten.detach.default](args = (%arg2_1,), kwargs = {})
    %split_with_sizes_1 : [num_users=2] = call_function[target=torch.ops.aten.split_with_sizes.default](args = (%detach_1, [%_local_scalar_dense, %_local_scalar_dense_1]), kwargs = {})
    %getitem_4 : [num_users=2] = call_function[target=operator.getitem](args = (%split_with_sizes_1, 0), kwargs = {})
    %getitem_5 : [num_users=0] = call_function[target=operator.getitem](args = (%split_with_sizes_1, 1), kwargs = {})
    %select_2 : [num_users=1] = call_function[target=torch.ops.aten.select.int](args = (%getitem_4, 0, 0), kwargs = {})
    %select_scatter : [num_users=1] = call_function[target=torch.ops.aten.select_scatter.default](args = (%select_2, %copy, 0, 0), kwargs = {})
    %select_scatter_1 : [num_users=1] = call_function[target=torch.ops.aten.select_scatter.default](args = (%getitem_4, %select_scatter, 0, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=operator.add](args = (0, %_local_scalar_dense), kwargs = {})
    %slice_scatter : [num_users=2] = call_function[target=torch.ops.aten.slice_scatter.default](args = (%detach_1, %select_scatter_1, 0, 0, %add), kwargs = {})
    %detach_2 : [num_users=1] = call_function[target=torch.ops.aten.detach.default](args = (%slice_scatter,), kwargs = {})
    %split_with_sizes_2 : [num_users=2] = call_function[target=torch.ops.aten.split_with_sizes.default](args = (%detach_2, [%_local_scalar_dense, %_local_scalar_dense_1]), kwargs = {})
    %getitem_6 : [num_users=1] = call_function[target=operator.getitem](args = (%split_with_sizes_2, 0), kwargs = {})
    %getitem_7 : [num_users=0] = call_function[target=operator.getitem](args = (%split_with_sizes_2, 1), kwargs = {})
    %select_3 : [num_users=1] = call_function[target=torch.ops.aten.select.int](args = (%getitem_6, 0, 0), kwargs = {})
    %select_4 : [num_users=0] = call_function[target=torch.ops.aten.select.int](args = (%select_3, 0, 0), kwargs = {})
    %detach_3 : [num_users=1] = call_function[target=torch.ops.aten.detach.default](args = (%arg2_1,), kwargs = {})
    %detach_4 : [num_users=1] = call_function[target=torch.ops.aten.detach.default](args = (%slice_scatter,), kwargs = {})
    %copy_ : [num_users=0] = call_function[target=torch.ops.aten.copy_.default](args = (%detach_3, %detach_4), kwargs = {})
    return (arg2_1, arg3_1, arg4_1)
```

IvanKobzarev replied:
I've updated the PR.

The logic now for this check - "We allow input mutations only, and mutations on input.detach()"

```
            if n.target is torch.ops.aten.copy_.default:
                    mutation_dst = n.args[0]
                    if mutation_dst.target == torch.ops.aten.detach.default:
                        mutation_dst = mutation_dst.args[0]

                    assert (
                        mutation_dst in placeholders ... 
```

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -671,8 +662,16 @@ def view_avoid_dupes_with_primals(t):
         # Anything that aliases (inputs returned in the fw due to metadata mutations, or outputs that alias inputs/intermediates)
         # are *regenerated* later, and not used directly in the autograd graph
+        def _plain_fake_tensor_like_subclass(x):
+            t = torch.empty(x.shape, dtype=x.dtype, device=x.device)
```
this is going to allocate actual cuda memory at compile time, which we probably don't want. You probably want this instead:
```
with detect_fake_mode():
    Return torch.empty(x.shape, dtype=x.dtype, device=x.device, layout=x.layout)
```

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -671,8 +662,16 @@ def view_avoid_dupes_with_primals(t):
         f_input_tangents = [
-            inp
+            # For subclass inputs that are mutated out of graph tangent will be
+            # plain zeros tensor like subclass input.
```
hmm I think this comment doesn't really give the full picture. I would probably say something like:
```
# Generally when creating tangents to trace with, we assume that tangents will have
# the same subclass-ness as their forward outs
# however: for tangents that correspond to input mutations, in practice it is more likely
# that these tangents will be plain tensors of zeros at runtime, so we tweak our guess
# to assume that the these tangents should always be plaint tensors.
# Example:
#  def f(x):
        x.mul_(2)
        return x + 1
#  out = f(x)
#  out.sum().backward()
# In the above code, we will have a tangent "x_updated_tangent",
# which will be a plain tensor of zeros, *unless* x is used in some compute after executing f
```

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1910,15 +1910,38 @@ def _backward_prologue(ctx, *flat_args):
+                    )
+
+                    flat_processed_tangents: List[torch.Tensor] = []
+                    if _num_mutated_inp_runtime > 0:
```
hmm what is the new logic here for?

IvanKobzarev replied:
Thanks, good catch. This is artifact from previous version, cleaning it.

bdhirsh commented on torch/_inductor/lowering.py:
```diff
@@ -920,7 +920,10 @@ def squeeze(x, dim=None):
-        if not (d in dims and V.graph.sizevars.evaluate_expr(sympy.Eq(s, 1))):
+        if not (
+            d in dims
+            and V.graph.sizevars.evaluate_expr(sympy.Eq(s, 1, size_oblivious=True))
```
did you need `size_oblivious=True` as part of this PR?

IvanKobzarev replied:
Yes, it is needed for symbolic shapes processing of the testcase for NJT

bdhirsh commented on torch/_meta_registrations.py:
```diff
@@ -4781,7 +4783,9 @@ def meta_select(self, dim, index):
     size = self.size(dim)
 
     torch._check_index(
-        not (-index > size or index >= size),
```
same with this?

IvanKobzarev replied:
yes, it is needed for NJT symbolic shapes processing for the test_case. There is njt index_put which does split underneath.

bdhirsh commented on torch/nested/_internal/ops.py:
```diff
@@ -929,20 +930,49 @@ def unbind_int(func, *args, **kwargs):
     lengths = inp.lengths()
     ragged_idx = inp._ragged_idx
 
+    def _torch_check(_lengths: List[int], _offsets: Optional[List[int]] = None):
```
cc @jbschlosser @soulitzer for the NJT changes

jbschlosser replied:
I'm good with these. It's also super exciting that we can use `unbind()` now within compile without graph breaks, as long as `torch._dynamo.config.capture_scalar_outputs = True` is set.

one minor nit would be to add some small amount of documentation describing the reason for this logic. (i.e. to provide torch.compile() with the information it's unable to obtain itself regarding the relationship among unbacked SymInts that result from the data-dependency on offsets)

IvanKobzarev replied:
Yes, _torch_check() info was needed to bypass symbolic shapes during split/unbind. Will add a comment.

bdhirsh commented on torch/nested/_internal/ops.py:
```diff
@@ -628,7 +629,7 @@ def copy_default(func, *args, **kwargs):
             "copy_ only supports Nested Tensors that have same size and the exact same offset tensor."
         )
-    inp.values().copy_(src.values())
+    inp._values.copy_(src._values)
```
Can you add a comment here on why it's actually important for us to use `_values` instead of `values()` here

bdhirsh reviewed (commented):
Looks mostly good, left a few comments

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutaiton in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -8558,6 +8549,33 @@ def f(*args, **kwargs):
                     self.assertEqualNoncontigAware(grads_compile, grads_ref)
 
+    @torch._dynamo.config.patch(capture_dynamic_output_shape_ops=True)
+    @torch._dynamo.config.patch(capture_scalar_outputs=True)
```
just to clarify: this flag is required for things to work? This is because the index_put implementation requires access to the offsets data, right?

IvanKobzarev replied:
Right. Index_select/Index_put operations are doing split underneath that needs access to offsets.item(), which only works with `capture_scalar_outputs=True`

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -8488,20 +8488,11 @@ def in_f(*args, **kwargs):
-                        # see https://github.com/pytorch/pytorch/issues/106456
-                        with self.assertRaisesRegex(
-                            RuntimeError,
-                            "Mutations on non-contiguous inputs are currently not "
-                            "allowed on tensor subclasses",
-                        ):
-                            compiled_in_f(sample.input, *sample.args, **sample.kwargs)
+                        self.assertEqual(sample.input, out_ref)
```
nice cleanup!

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -8558,6 +8549,33 @@ def f(*args, **kwargs):
+            return torch.nested.nested_tensor_from_jagged(
+                v,
+                offsets=torch.tensor([0, 2, 3]),
+                # lengths=torch.tensor([2, 1]),
```
nit: remove commented-out code or utilize it?

soulitzer commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -671,8 +661,45 @@ def view_avoid_dupes_with_primals(t):
+            # the same subclass-ness as their forward outs
+            # however: for tangents that correspond to input mutations, in practice it is more likely
+            # that these tangents will be plain tensors of zeros at runtime, so we tweak our guess
+            # to assume that these tangents should always be plaint tensors.
```
Working through this example to try to understand this better:
```
def f(x):
    x.mul_(2)
    return x + 1
```

```
def inner(x):
    x_modified = x * 2
    return x, x_modified, x_modified + 1
    
def outer(x):
    x_input, x_modified, out = inner(x)
    x_input.copy_(x_modified)
    return out
```
Some noob questions (I might be missing some crucial AOTAutograd context):

In this case the tangent that corresponds to the `x` is zeros/undefined right? Is that what you mean by "tangents that correspond to input mutations"? (I was confused by the naming of x_updated_tangent in the comment, because x_updated should have non-zero gradient right)

Is the reason why the tensor is assumed to be a plain tensor because when custom autograd Function receives None, it materializes a zeros in the shape of the output shape and does not reconstruct the subclass? (If this is the case, that's not actually true for NestedTensor which will actually create a subclass in this case).

soulitzer replied:
> If this is the case, that's not actually true for NestedTensor which will actually create a subclass in this case

I might be missing something here, because if this is true, then we would've had to use `disable_guess_zero_tangent_for_mutated_input_subclass` to special case NestedTensor...

bdhirsh replied:
> it materializes a zeros in the shape of the output shape and does not reconstruct the subclass? (If this is the case, that's not actually true for NestedTensor which will actually create a subclass in this case).

oh hmm... @soulitzer where does the logic for that live? (is there some special-casing of NJT in the autograd engine?)

soulitzer replied:
Relevant PR here. https://github.com/pytorch/pytorch/pull/136875
maybe relevant issue https://github.com/pytorch/pytorch/issues/133644

Not in engine but in custom autograd Function. For nested tensor specifically it is important that the subclassness of a tensor matches the subclassness of the tangent because the subclassesness is what defines the tensor's shape, and autograd requires the shape to be equivalent.

bdhirsh replied:
hmm @IvanKobzarev it sounds like we should special-case NJT in that tangent-guessing decision. Any chance you're able to repro the error Jeffrey is talking about? I think you should be able to repro with code like this:
```
@torch.compile
def f(njt):
    njt.mul_(2)
    return njt + 1

# do an input mutation on an NJT that requires grad
x = NestedTensor(..., requires_grad=True)
out = f(x)
# don't use x anywhere after out is computed
# According to Jeffrey, NJT is special-cased so that the `x_updated_tangent` will **not** be a plain ttensor here
out.sum().backward()
```

IvanKobzarev replied:
@bdhirsh @soulitzer 
I tried this variant, but for NJT this fails in eager mode, as view of Leaf-variable is mutated:
```
        def fn(x):
            x.mul_(2)
            return x + 1

        def _inp():
            v = torch.zeros(32, 3, requires_grad=True)
            return torch.nested.nested_tensor_from_jagged(
                v,
                offsets=torch.tensor([0, 2, 3]),
            )
        
        ref_x = _inp()
        ref_y = fn(ref_x)
        ref_y.sum().backward()
```
```
Traceback (most recent call last):
  File "/data/users/ivankobzarev/b/pytorch/torch/testing/_internal/common_utils.py", line 3099, in wrapper
    method(*args, **kwargs)
  File "/data/users/ivankobzarev/b/pytorch/torch/testing/_internal/common_utils.py", line 3099, in wrapper
    method(*args, **kwargs)
  File "/data/users/ivankobzarev/b/pytorch/torch/testing/_internal/common_device_type.py", line 480, in instantiated_test
    raise rte
  File "/data/users/ivankobzarev/b/pytorch/torch/testing/_internal/common_device_type.py", line 460, in instantiated_test
    result = test(self, **param_kwargs)
  File "/data/users/ivankobzarev/b/pytorch/test/test_nestedtensor.py", line 8592, in test_nested_tensor_input_mutation_backward
    ref_y = fn(ref_x)
  File "/data/users/ivankobzarev/b/pytorch/test/test_nestedtensor.py", line 8580, in fn
    x.mul_(2)
  File "/data/users/ivankobzarev/b/pytorch/torch/nested/_internal/nested_tensor.py", line 353, in __torch_function__
    return func(*args, **kwargs)
RuntimeError: a view of a leaf Variable that requires grad is being used in an in-place operation.
```

IvanKobzarev replied:
From offline discussion:

NJT has a special handling in Autograd, that even if Autograd wants to create zero tensor, for NJT it will call zeros_like() and the tangent is guranteed to be NJT even if it is not used. (see torch/csrc/autograd/python_function.cpp use_zeros_like)

Adding special case to not guess plain_zero_tensor for NJT.

Adding test for this case in test_nestedtensor.py

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutation in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutation in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutation in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

bdhirsh commented on torch/_functorch/config.py:
```diff
@@ -205,6 +205,10 @@ def remote_autograd_cache_default() -> Optional[bool]:
 strict_autograd_cache = False
 
+# See Note [AOTAutograd Tangent Subclassness for mutated inputs]
+# TODO(ivankobzarev): Remove this config, being able to deduce it compile time.
```
idk if we're ever actually going to be able to deduce this at compile time, unless we actually commit to the "don't specialize tangents until backward runs" work :/

bdhirsh reviewed (approved):
nice!

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[aotd] non-contiguous NestedTensor mutation in compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Allow mutations mutations for subclasses that are non-contiguous.

Changes:

Removing assert in collect_metadata_analysis

Main requested testcase:
Compilation of NJT.index_put()

Adding test in test_nestedtensor.py, that compiles NJT.index_put()

It is  decomposed to NJT split,unbind, which  needed additional `torch._check`, `torch._check_is_size` for NJT.unbind()  and guard_size_oblivious() usage in _meta_registrations and _inductor/lowering.py.


Special case:
If tangent is mutated outside of the graph, it does not participate in backward graph. Autograd in this case will set this tangent to zeros tensor.

We handle it separately in CompiledFunction.backward: not doing any processing for this tangent and broadcast to number of expected subclass unwrapped arguments.

disabling for dynamo 2 tests:
1/ For nested tensor - symbolic shapes issue on nested_tensor index operation that does splits [0, 0, 0] - there is a failure with "pending unbacked symints". This PR does not add more .tolist()/item() ops than it was before.

2/ As we do not fail with exception in collect_metadata_analysis new paths for dynamo started working and it started failing with smth strange that set_ in storage_offset (because of test for views) handling updates storage "cpu" -> "meta"




[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge
